### PR TITLE
[Structural] Perturb Geometry Process

### PR DESCRIPTION
**Pull request by manuelmessmer**
**Description**
This PR adds a geometry perturbation process to the StructuralMechanicsApplication. With this process the user can perturb the initial geometry of a structure. This is especially useful in combination with the buckling solver in Kratos, as you can analyze
the buckling behaviour of imperfect geometries.
The process is based on random field theory. In a nutshell: A correlation matrix (size:NxN) for all nodes in the model part is assembled. The correlation value is based on the euclidean distance between the nodes. If nodes are close together -> large value, if further apart -> small value. The correlation matrix is decomposed and the eigenvectors are superposed to form the random field. Each eigenvector has a random number associated to it, such that an arbitrary amount of unique random fields can be derived from the same eigenvectors. 

**Changelog**
Adds:
- PerturbGeometrySparseProcess based on a sparse correlation matrix
- PerturbGeometrySubgridProcess based on a correlation matrix in a reduced space (coarser mesh)
- OMP_NodeSearch to find a reduced space with equidistant nodes.
- NodeConfigureForNodeSearch (Intersection function is based on euclidean distance among two nodes)
- Python interface for processes
- Unit test

**Additional info**
The so-called correlation length is a user defined parameter, which roughly defines the characteristic wavelength of the final perturbations -> see images below for small and large correlation length values. In both images the original geometry is a straight plate.
In theory the correlation matrix is a NxN dense matrix. To be able to apply this process also to larger models, there are two different implementations. The first one assembles an artificially sparse correlation matrix and the second approach solves the random field problem in a subgrid (coarse mesh) and then maps the final perturbation field back to the full domain.
small_correlation_length
large_correlation_length

### CODE REVIEW

**Comment by philbucher** [comment]
overall I am fine with the additions, but still I would like to speak to you about a couple of things in person in the next days, I think this will be easier than writing everything here :)

**Comment by manuelmessmer** [comment]
Hi @philbucher, I think I addressed all your comments in the last commits. I also switched to the parallel_utilities, where no private() is required.

File: applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.h]:
```suggestion
```

I suggest to remove the unused sections

same for other files

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]:
```suggestion
            for(std::size_t i = 0; i < results.size(); ++i){
```

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]: do you need the iterator loop?

```suggestion
    for (auto it_node = mrInitialModelPart.NodesBegin(); it_node != mrInitialModelPart.NodesEnd(); ++it_node){
```

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]: same

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]: same

**Review by philbucher — COMMENTED** [review]
minor comments otherwise looks ok

**Comment by manuelmessmer** [comment]
Hi @philbucher, is this okay now?

File: applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.cpp]: this should not be done (in fact this is only possible because of a very special smart pointer we use in Kratos)
Basically this messes up the memory, you cannot create a smart pointer from an object that is already managed by another smart pointer (more details in person)

I think you will have to change the function signature

(TBH I think the bins should not take a smart pointer as input but that is a different discussion)

File: applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/node_search_utility.cpp]:
```suggestion
    auto results_iterator = local_results.begin();
```

IMO a good opportunity to use `auto`

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]:
```suggestion
```

I would remove this since it is a local variable

Otherwise you have two memory related operations (those are expensive)

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp:124]: this can also be done using the Parallel Utilities once #7261 is merged

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp:74]: +1 for using ParallelUtilities

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp]: can be removed
```suggestion
```

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_base_utility.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_base_utility.h]: can you move this to the cpp please?
then you can also move some includes

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp]: this can also easily be parallalized with #7261 

(actually I added an example for this usecase)

**Review by philbucher — COMMENTED** [review]
sorry for the delay, I thought this was merged already

I made some final important comments, after that looks good to me

thanks again for the patience

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp:124]: Shall I wait until #7261 is merged or make the changes afterwards in a new PR?

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp
Comment (manuelmessmer) [reply applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp]: The problem here is that the inner for loop is of varying size.

Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_sparse_utility.cpp]: i meant the outter loop over the nodes

File: applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp
Comment (philbucher) [reply applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp:124]: no need, we can refactor later

**Review by philbucher — APPROVED** [review]
:+1:

**Comment by manuelmessmer** [comment]
@philbucher thanks for reviewing!! Then I will merge and close this and do the remaining changes in a separate PR once #7261 is merged?

**Pull request merged**
